modules/hal_rpi_pico/CMakeLists.txt

```diff
@@ -0,0 +1,108 @@
+if(CONFIG_HAS_RPI_PICO)
+	zephyr_library()
+
+	set(RP2_COMMON_DIR ${ZEPHYR_HAL_RPI_PICO_MODULE_DIR}/src/rp2_common)
+	set(RP2040_DIR ${ZEPHYR_HAL_RPI_PICO_MODULE_DIR}/src/rp2040)
+	set(COMMON_DIR ${ZEPHYR_HAL_RPI_PICO_MODULE_DIR}/src/common)
+	set(BOOT_STAGE_DIR ${RP2_COMMON_DIR}/boot_stage2)
+
+	# Compile the second stage bootloader as a separate executable
+
+	add_executable(boot_stage2)
```

Comment:
Indeed, this bootloader does not have anything Zephyr specific. The bootloader's goal is to set up the XIP code execution. The only reason it isn't flashed to the ROM by the manufacturer is that the bootloader varies with the flash chip being used.
Regarding the flash offset: the reason for the unconventional image relocation by truncating the first 256 bytes of the flash was that if the offset was given by dts partitions, it would require dts partitions to always be used. As I didn't want to force any partition layout, I preferred to just start the flash from 0x100 (that area isn't usable anyway). I did see this issue: https://github.com/zephyrproject-rtos/zephyr/issues/27471 which addresses this problem, but as it doesn't seem to have a solution yet I left it as it was. If you have any better solution I would love to use it.

Reply:
@tejlmand the RP2040 has ROM internally that sets up the machine and calls the Flash bootloader to do the rest. The flash bootloader's job is purely to enable code to run from the Flash chip (by initialising QSPI and XIP) for the specific flash on the board.

AFAIK the right thing to do if you want to use another bootloader (e.g. because it offers some sort of monitor prompt, or access to a filesystem) would be to chainload from the standard rpi code into the new code. The standard code really is trivially simple.

@yonsch is therefore doing the right thing IMO.

Reply:
the main problem with this approach is that several attempts to create custom multi-image builds in Zephyr has been rejected, because a common solution has been wanted on higher system level.

Allowing this now for the raspberry pico, when other requests has been turned down doesn't seem fair.

Instead I would like to get feedback on #40555 as to how that would fit with the need presented in this PR.
Also feel free to create dedicated use-cases for your particular need and link them to: https://github.com/zephyrproject-rtos/zephyr/issues/41640

Instead of n-custom solutions, I would very much like that we work on a common direction, so that the Zephyr build itself is kept free of any bootloader firmware, custom as well as MCUboot.

When building Zephyr with MCUBoot support, then MCUboot must also be build and flashed independently of Zephyr in order to have a working system.

In this case the second stage bootloader must be flashed.
But as long as you don't erase it, you can rebuild and reflash the Zephyr based app as often as you like.
So flashing of the specific second stage bootloader for the given flash would be a one-time operation, correct ?

Reply:
ref: https://github.com/yonsch/zephyr/pull/3

Reply:
Original comment: https://github.com/yonsch/zephyr/pull/3#issuecomment-1018087439

Answered here for others to chip in.
> One comment: this work is basically not board specific and there are at least 3 more boards in the queue waiting to be added** which will need the same cmake files at least. Can the new code be moved under the 'soc' tree instead, where it can be shared nicely?

Very good point.
Makes a lot of sense.

Currently we don't have `samples/soc`, but we could of course create one.
But we can also consider if we should instead create a folder for those sample that are not creating a full Zephyr based app.

I think we should consider among the following possible options:
1.  `samples/boards/rpi_pico/boot_stage2`: Current solution, very board specific.
2. `samples/soc/<soc>/boot_stage2`: Folder to be created, we currently don't have soc specific samples, but makes a lot of sense to have one.
3. [samples/modules/](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/modules)hal_rpi_pico/boot_stage2 : Source is picked up from `hal_rpi_pico` Zephyr module, so this location relates well to where boot source is located.
4. `samples/non-zephyr/boot_stage2`: New folder for non-zephyr based samples.

Of the options above, I have a preferency for 4. because it makes it clear that this bootloader in not a regular Zephyr sample.
But option 3. is my runner-up favorite, as it closely relates to the Zephyr module where the source files are located.

> 
> 2 from adafruit, one solderparty; there are many others out there.

Great to hear,

Reply:
Hi, my initial comment was not specifically aimed at 'samples' but at the board in general. My understanding is that you have a 3 level hierarchy: arch, soc, and board, where arch is 'arm', 'riscv, etc, and soc is the chip/device providing most of the cpu, mem and peripherals, and board is the soc as fitted on a board, including any board-installed peripherals etc.

In this case the triple is (or should be) (arm, rp2040, rpi_pico).

A lot of stuff in this series of commits is being applied to 'board' and should in fact be targetted at 'soc', because it would apply equally to any board based on the rp2040, and in particular the method of avoiding use of a boot stage2 (which is needed because of the rom on the rp2040).

Should the samples also be reliant only on the features of the soc then of course they should be soc-specific samples but that wasn't really my point.